src/pedersen_hash.rs

```diff
@@ -78,8 +79,23 @@ pub fn pedersen_hash<E, I>(
             break;
         }
 
-        let mut tmp = generators.next().expect("we don't have enough generators").clone();
-        tmp = tmp.mul(acc, params);
+        let mut table: &[Vec<edwards::Point<E, _>>] = &generators.next().expect("we don't have enough generators");
+        let window = JubjubBls12::pedersen_hash_exp_window_size();
+        let window_mask = (1 << window) - 1;
+
+        let mut acc = acc.into_repr();
+        
+        let mut tmp = edwards::Point::zero();
+
+        while !acc.is_zero() {
+            let i = (acc.as_ref()[0] & window_mask) as usize;
+
+            tmp = tmp.add(&table[0][i], params);
```

Comment:
Maybe if each table were an ``[edwards::point<E, _>; 256]``, a Sufficiently Smart Compiler could optimize out the bounds check? I don't know whether the Rust compiler is that smart (also, this is a microoptimization that's probably insignificant relative to the field multiplications).